In [1]:
import pandas as pd

# Load the datasets
diabetic_data_path = r"D:\7333\Case Study 2\diabetic_data.csv"
ids_mapping_path = r"D:\7333\Case Study 2\IDs_mapping.csv"


diabetic_data = pd.read_csv(diabetic_data_path)
ids_mapping = pd.read_csv(ids_mapping_path)

# Display basic information about the datasets
diabetic_data_info = diabetic_data.info()
ids_mapping_info = ids_mapping.info()

# Display the first few rows of each dataset
diabetic_data_head = diabetic_data.head()
ids_mapping_head = ids_mapping.head()

diabetic_data_info, diabetic_data_head, ids_mapping_info, ids_mapping_head


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

(None,
    encounter_id  patient_nbr             race  gender      age weight  \
 0       2278392      8222157        Caucasian  Female   [0-10)      ?   
 1        149190     55629189        Caucasian  Female  [10-20)      ?   
 2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
 3        500364     82442376        Caucasian    Male  [30-40)      ?   
 4         16680     42519267        Caucasian    Male  [40-50)      ?   
 
    admission_type_id  discharge_disposition_id  admission_source_id  \
 0                  6                        25                    1   
 1                  1                         1                    7   
 2                  1                         1                    7   
 3                  1                         1                    7   
 4                  1                         1                    7   
 
    time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
 0                 1  ...          No      



### Summary of the Data:
1. **Diabetic Data (`diabetic_data.csv`):**
   - Contains **101,766** rows and **50** columns.
   - Includes patient demographics, medical history, medications, and hospitalization details.
   - The target variable is **"readmitted"**, which has three categories:
     - **NO** (not readmitted)
     - **>30** (readmitted after 30 days)
     - **<30** (readmitted within 30 days)
   - Several categorical features contain missing values represented by `?`.
2. **IDs Mapping (`IDs_mapping.csv`):**
   - Contains **67 rows** and **2 columns**.
   - Maps **admission type IDs** to their descriptions.

### Next Steps:
1. **Data Cleaning & Preprocessing**
   - Replace missing values (`?`) with appropriate imputation techniques.
   - Convert categorical variables into numerical representations.
   - Map admission type IDs to their descriptions.
2. **Feature Engineering & Selection**
   - Identify the most relevant variables affecting readmission.
3. **Modeling: Logistic Regression**
   - Build a model to classify the readmission categories.
   - Perform hyperparameter tuning.
4. **Evaluation**
   - Report confusion matrices, classification reports, and AUC-ROC scores.


In [4]:
import pandas as pd

# Load the datasets

diabetic_data_path = r"D:\7333\Case Study 2\diabetic_data.csv"
ids_mapping_path = r"D:\7333\Case Study 2\IDs_mapping.csv"

diabetic_data = pd.read_csv(diabetic_data_path)
ids_mapping = pd.read_csv(ids_mapping_path)

# Display basic information about the datasets
diabetic_data_info = diabetic_data.info()
ids_mapping_info = ids_mapping.info()

# Display the first few rows of each dataset
diabetic_data_head = diabetic_data.head()
ids_mapping_head = ids_mapping.head()

diabetic_data_info, diabetic_data_head, ids_mapping_info, ids_mapping_head


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

(None,
    encounter_id  patient_nbr             race  gender      age weight  \
 0       2278392      8222157        Caucasian  Female   [0-10)      ?   
 1        149190     55629189        Caucasian  Female  [10-20)      ?   
 2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
 3        500364     82442376        Caucasian    Male  [30-40)      ?   
 4         16680     42519267        Caucasian    Male  [40-50)      ?   
 
    admission_type_id  discharge_disposition_id  admission_source_id  \
 0                  6                        25                    1   
 1                  1                         1                    7   
 2                  1                         1                    7   
 3                  1                         1                    7   
 4                  1                         1                    7   
 
    time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
 0                 1  ...          No      

In [2]:
import numpy as np

# Replace '?' with NaN for proper handling of missing values
diabetic_data.replace('?', np.nan, inplace=True)

# Check for missing values again
missing_values = diabetic_data.isnull().sum()

# Display columns with missing values
missing_values[missing_values > 0]


race                  2273
weight               98569
payer_code           40256
medical_specialty    49949
diag_1                  21
diag_2                 358
diag_3                1423
max_glu_serum        96420
A1Cresult            84748
dtype: int64

### Missing Data Overview:
- **Race**: 2,273 missing values
- **Weight**: 98,569 missing values (almost entirely missing, likely to be dropped)
- **Payer Code**: 40,256 missing values
- **Medical Specialty**: 49,949 missing values (about 50% missing)
- **Diagnosis Codes (diag_1, diag_2, diag_3)**: Some missing values

### Imputation Strategy:
- **Drop `weight`** since it's missing in nearly all cases.
- **Drop `payer_code` and `medical_specialty`** due to a high number of missing values.
- **Impute `race` with "Unknown"** since it is categorical.
- **Impute `diag_1`, `diag_2`, and `diag_3` with "Missing"** as they are categorical variables.

In [5]:
import numpy as np

# Replace '?' with NaN for proper handling of missing values
diabetic_data.replace('?', np.nan, inplace=True)

# Check for missing values again
missing_values = diabetic_data.isnull().sum()

# Display columns with missing values
missing_values[missing_values > 0]


race                  2273
weight               98569
payer_code           40256
medical_specialty    49949
diag_1                  21
diag_2                 358
diag_3                1423
max_glu_serum        96420
A1Cresult            84748
dtype: int64

In [8]:
print(diabetic_data.columns)


Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')


In [12]:
# Drop columns with excessive missing values
diabetic_data.drop(columns=['weight', 'payer_code', 'medical_specialty'], inplace=True)

# Impute missing categorical values
diabetic_data['race'].fillna('Unknown', inplace=True)
diabetic_data['diag_1'].fillna('Missing', inplace=True)
diabetic_data['diag_2'].fillna('Missing', inplace=True)
diabetic_data['diag_3'].fillna('Missing', inplace=True)

# Verify that no missing values remain
missing_values_after = diabetic_data.isnull().sum()
missing_values_after[missing_values_after > 0]


In [9]:
# Impute missing categorical values
diabetic_data = diabetic_data.assign(
    race=diabetic_data['race'].fillna('Unknown'),
    diag_1=diabetic_data['diag_1'].fillna('Missing'),
    diag_2=diabetic_data['diag_2'].fillna('Missing'),
    diag_3=diabetic_data['diag_3'].fillna('Missing')
)

# Verify that no missing values remain
missing_values_after = diabetic_data.isnull().sum()
print(missing_values_after[missing_values_after > 0])


max_glu_serum    96420
A1Cresult        84748
dtype: int64


In [11]:
#  Keeping cvhained approach 
 
diabetic_data.loc[:, 'race'] = diabetic_data['race'].fillna('Unknown')
diabetic_data.loc[:, 'diag_1'] = diabetic_data['diag_1'].fillna('Missing')
diabetic_data.loc[:, 'diag_2'] = diabetic_data['diag_2'].fillna('Missing')
diabetic_data.loc[:, 'diag_3'] = diabetic_data['diag_3'].fillna('Missing')







Handle missign values: encode categorical variables and prepare the dataset for logistic regression by transforming the target variable  `readmitted` target variable is now numerical (0: NO, 1: >30, 2: <30). 

Split the data into training and testing sets, then build logistic regression classifiers for predicting hospital readmission.

 split into training and testing sets:
- **Training Set**: 81,412 samples, 44 features.
- **Testing Set**: 20,354 samples, 44 features.

-   Ntext train logistic regression classifiers and evaluate their performance.
### Model Evaluation:
**Confusion Matrix:**
\[
\begin{bmatrix}
9898 & 1063 & 12 \\
5406 & 1680 & 23 \\
1604 & 642 & 26
\end{bmatrix}
\]
- **True Negatives (NO predicted as NO)**: 9,898
- **False Positives (NO predicted as >30 or <30)**: 1,063 + 12
- **False Negatives (Actual >30 or <30 predicted as NO)**: 5,406 + 1,604
- **Correct Predictions for >30**: 1,680
- **Correct Predictions for <30**: 26

**Classification Report:**
- **Overall Accuracy**: 57%
- **Precision, Recall, and F1-score per Class:**
  - **NO** (Not Readmitted): Precision = 0.59, Recall = 0.90
  - **>30** (Readmitted after 30 days): Precision = 0.50, Recall = 0.24
  - **<30** (Readmitted within 30 days): Precision = 0.43, Recall = 0.01

### Observations:
- The model performs well in predicting **NO readmission** cases but struggles with readmission cases (`>30` and `<30`).
- **Recall for `<30` readmission is extremely low (1%)**, meaning the model fails to identify most patients readmitted within 30 days.
- **Precision for `>30` is 50%**, but its recall is low (24%), indicating poor balance.

### Next Steps:
1. **Feature Importance Analysis** to determine the top 5 variables influencing predictions.
2. **Hyperparameter Tuning** to improve recall on `<30` cases.
3. **Consider Alternative Models** (Random Forest, XGBoost) to handle class imbalance.
4. **Grid Model**


### Top 5 Most Important Features:
1. **Number of inpatient visits** (0.226) – Strongest predictor of readmission.
2. **Number of emergency visits** (0.087) – Higher emergency visits increase the likelihood of readmission.
3. **Number of diagnoses** (0.070) – More diagnoses correlate with more complex medical conditions and higher readmission risk.
4. **Discharge disposition ID** (0.069) – Likely relates to where the patient is discharged (e.g., home, hospice, rehab).
5. **Diabetes medication use** (0.048) – Indicates whether diabetes management plays a role in readmission.

### Key Takeaways:
- **Number of inpatient stays dominates** as the most significant feature.
- **Emergency visits and multiple diagnoses** contribute significantly to predicting readmission.
- **Discharge disposition and diabetes medication** suggest patient care decisions influence readmission risk.

Next hyperparameter tuning to improve recall, especially for `<30` readmissions.


